### ISSUE TITLE
Whitespace and the annotation module

### ISSUE THREAD

**Issue by michmech** [oasis-tcs/lexidma#84]
(Submitted by John McCrae)

I have been thinking and I
am still not sure about our rule for whitespace in the elements. In
particular, in the converter I have been developing I am having problems
because we cannot apply pretty printing (indenting) to an XML file
without changing the content in the model. Further, I think the rules
are unintuitive and many will add whitespace and create unfortunate
errors. Instead I propose that we adopt the HTML methodology as
described here:

https://infra.spec.whatwg.org/#strip-newlines

In this case, before processing the content of any text carrying
element, we will first remove all new lines ('\n', '\r'), delete all
trailing and leading whitespace and replace all remaining blocks of
ASCII whitespace with a single space.

I would also make a model change, replacing all references to 'non-empty
string' with a 'normalised string'. This means a string that contains no
new lines, does not start or end with a whitespace, contains no block of
ASCII whitespace more than a single space and is non-empty. This ensures
that other serializations (JSON, RDF) cannot generate content that
cannot be represented in XML.

I do worry that this does not really cover Chinese, Japanese (and maybe
Thai/Lao), as the whitespace rules for HTML are more complex in Unicode,
but I think that this can probably be worked around by lexicographers
working in these languages. We can add a note to the spec for these
languages.

**Comment by michmech** [comment]
I agree with John’s suggestion.

At the XML serialization level, we need to make sure that we don’t accidentally cause the removal of white-space only text nodes between two instances of inline markup, eg. in a headword like headword “give sb. sth.”  between the “sb.” and the “sth.”:

```
<headword>give <placeholderMarker>sb.</placeholderMarker> <placeholderMarker>sth.</placeholderMarker></headword>
```

**Comment by jmccrae** [comment]
Yes, this proposal would mean that whitespace is only completely remove after an opening tag and before a closing tag. The example above has whitespace after a closing tag (`</placeholdMarker>`) and before an opening tag (`<placeholderMarker>`) so it is preserved.

**Comment by michmech** [comment]
I nave now implemented this proposal.

At the model level, I have changed all occurrences of 'non-empty string' to 'normalised string' and I have explained (in section 1.3.2 'Properties') what that means. Also, I have made sure that *all* literal properties of *all* objects say what type they are (normalised string, number etc.) and I have updated the list of literal types in section 1.3.2.

At the XML serialisation level, I have removed the (non-normative) prose we had about whitespace and replaced it with a new (normative) section where we explain that whitespace is _always_ insignificant _except_ inside elements that are allowed to have text content, where the text content should be treated as 'normalised string' as defined in section 1.3.2.
